rllib/algorithms/algorithm.py

```diff
@@ -1595,35 +1599,98 @@ def add_policy(
+                policy_state=policy_state,
+                policy_mapping_fn=policy_mapping_fn,
+                policies_to_train=list(policies_to_train)
+                if policies_to_train
+                else None,
+            )
+        # Policy instance provided: Create clones of this very policy on the different
+        # workers (copy all its properties here for the calls to add_policy on the
+        # remote workers).
+        else:
+            kwargs = dict(
+                policy_id=policy_id,
+                policy_cls=type(policy),
+                observation_space=policy.observation_space,
+                action_space=policy.action_space,
+                config=policy.config,
+                policy_state=policy.get_state(),
+                policy_mapping_fn=policy_mapping_fn,
+                policies_to_train=list(policies_to_train)
+                if policies_to_train
+                else None,
+            )
 
         def fn(worker: RolloutWorker):
             # `foreach_worker` function: Adds the policy the the worker (and
             # maybe changes its policy_mapping_fn - if provided here).
             worker.add_policy(**kwargs)
 
+        # Workers to add the policy to are given as an explicit list.
         if workers is not None:
             ray_gets = []
             for worker in workers:
+                # A remote worker (ray actor).
                 if isinstance(worker, ActorHandle):
                     ray_gets.append(worker.add_policy.remote(**kwargs))
+                # (Local) RolloutWorker instance.
                 else:
-                    fn(worker)
+                    if policy is not None:
+                        worker.add_policy(
+                            policy_id=policy_id,
+                            policy=policy,
+                            policies_to_train=policies_to_train,
+                            policy_mapping_fn=policy_mapping_fn,
+                        )
+                    else:
+                        fn(worker)
             ray.get(ray_gets)
+        # Add to all RolloutWorkers within `self.workers`.
         else:
+            # Policy is provided as an instance -> Add this very instance to local
+            # worker.
+            if policy is not None:
+                local_worker.add_policy(
+                    policy_id=policy_id,
+                    policy=policy,
+                    policies_to_train=policies_to_train,
+                    policy_mapping_fn=policy_mapping_fn,
+                )
+                # Then add a new instance to each remote worker.
+                ray.get([w.apply.remote(fn) for w in self.workers.remote_workers()])
             # Run foreach_worker fn on all workers.
-            self.workers.foreach_worker(fn)
+            else:
+                self.workers.foreach_worker(fn)
 
         # Update evaluation workers, if necessary.
         if evaluation_workers and self.evaluation_workers is not None:
```

Comment:
one thing that feels a bit weird is how come we can handle this for eval_workers with 1 line of code, but we need this many if_elses for rollout workers.
I wonder if we can do something similar for rollout workers, basically ``self.workers.foreach_worker(fn)``,
and WorkerSet.foreach_worker() will apply the fn locally on the local worker, or apply it remotely on all the remote workers.
that way, the logics about individual workers are capsulated behind the WorkerSet abstraction.
does that work??

Reply:
Eval workers don't have a local worker :)

For the (only one!) local worker, we should insert the policy directly into its policy_map, no re-creation of a new instance is required. That's the whole point of this PR, I guess.

The `foreach_worker` utility is actually fine (handles local worker properly) and has nothing to do with this.

Reply:
Cleaned up a little (removed the helper function entirely, not needed).

Reply:
wait, eval workers can have local worker too??
I often set evaluation_num_workers=0 for OPE, since we do OPE on trainer node anyways.
that will cause the evaluation_workers to use only a local worker?

also, one thing I am always a bit confused, if we already have a WorkerSet abstraction, why should Algorithm still manipulate individual local and remote workers itself. feel like it's better to have WorkerSet handle the underlying details?

for the specific case, I wonder if we should simply do something like:
```
rollout_workers.local_worker().add_policy(policy=policy)
rollout_workers.remote_worker().add_policy(policy_cls=type(policy), policy_state=...)
evaluation_workers.local_worker().add_policy(policy_cls=type(policy), policy_state=...)
evaluation_workers.remote_worker().add_policy(policy_cls=type(policy), policy_state=...)
```
then, the policy is only claimed by the local rollout worker.

Reply:
The evaluation worker set only has a local worker if `evaluation_num_workers=0`. Otherwise, we'll skip generating it.

If you do: Algorithm.evaluate(), it will:
* first try to use the evaluation worker set (be it with local worker (evaluation_num_workers=0) or without local worker).
* then, if there is NO evaluation worker set at all, use the regular local worker. Note that this only works, if that local worker has an env (by default, we don't create one on the regular local worker)

Reply:
I think you are right and we should create a `WorkerSet.add/remove_policy` API. Then we can move all the code that's currently in `Algorithm.add_policy` into the WorkerSet and in the algo, simply do:

```
def add_policy(...):
  self.workers.add_policy()
  self.evaluation_workers.add_policy()
  return
```

Reply:
done